src/goto-instrument/cover.cpp

```diff
@@ -11,1687 +11,196 @@ Date: May 2016
-            (is_decision && is_condition)?"decision/condition":
-            is_decision?"decision":"condition";
-
-          std::string p_string=from_expr(ns, "", p);
-
-          std::string comment_t=description+" `"+p_string+"' true";
-          const irep_idt function=i_it->function;
-          goto_program.insert_before_swap(i_it);
-          // i_it->make_assertion(p);
-          i_it->make_assertion(not_exprt(p));
-          i_it->source_location=source_location;
-          i_it->source_location.set_comment(comment_t);
-          i_it->source_location.set(ID_coverage_criterion, coverage_criterion);
-          i_it->source_location.set_property_class(property_class);
-          i_it->source_location.set_function(function);
-          i_it->function=function;
-
-          std::string comment_f=description+" `"+p_string+"' false";
-          goto_program.insert_before_swap(i_it);
-          // i_it->make_assertion(not_exprt(p));
-          i_it->make_assertion(p);
-          i_it->source_location=source_location;
-          i_it->source_location.set_comment(comment_f);
-          i_it->source_location.set(ID_coverage_criterion, coverage_criterion);
-          i_it->source_location.set_property_class(property_class);
-          i_it->source_location.set_function(function);
-          i_it->function=function;
-        }
-
-        std::set<exprt> controlling;
-        // controlling=collect_mcdc_controlling(decisions);
-        controlling=collect_mcdc_controlling_nested(decisions);
-        remove_repetition(controlling);
-        // for now, we restrict to the case of a single ''decision'';
-        // however, this is not true, e.g., ''? :'' operator.
-        INVARIANT(!decisions.empty(), "There must be at least one decision");
-        minimize_mcdc_controlling(controlling, *decisions.begin());
-
-        for(const auto &p : controlling)
-        {
-          std::string p_string=from_expr(ns, "", p);
-
-          std::string description=
-            "MC/DC independence condition `"+p_string+"'";
-
-          const irep_idt function=i_it->function;
-          goto_program.insert_before_swap(i_it);
-          i_it->make_assertion(not_exprt(p));
-          // i_it->make_assertion(p);
-          i_it->source_location=source_location;
-          i_it->source_location.set_comment(description);
-          i_it->source_location.set(ID_coverage_criterion, coverage_criterion);
-          i_it->source_location.set_property_class(property_class);
-          i_it->source_location.set_function(function);
-          i_it->function=function;
-        }
-
-        for(std::size_t i=0; i<both.size()*2+controlling.size(); i++)
-          i_it++;
-      }
-      break;
-
-    case coverage_criteriont::PATH:
-      if(i_it->is_assert())
-        i_it->make_skip();
-      break;
-
-    default:
-      {
-      }
-    }
+    std::stringstream s;
+    s << "unknown coverage criterion " << '\'' << criterion_string << '\'';
+    throw s.str();
```

Comment:
This feels a bit wrong as I'd want a user error to give a meaningful error message, not throw an exception.